### ISSUE TITLE
AbstractNodeQueue potential GC leak - many objects get tenured if we have "bursts" of messages

### ISSUE THREAD

**Issue by dom54** [akka/akka#17547]
Akka: 2.3.11
Scala: 2.11.4

I have noticed that when using the SingleConsumerOnlyUnboundedMailbox there are excessive amounts of garbage created given a certain frequency of messages.

I have traced the issue to the AbstractNodeQueue - if you have many actors (say, 500,000) and therefore many queues, **the last node created for the last message sent to each actor will become the tail, and therefore if no new messages are received for a period of a few minutes (or a few minor GCs), will be copied and eventually tenured - for 500,000 actors thats 500,000 node objects**

```
public final T poll() {
        final Node<T> next = peekNode();
        if (next == null) return null;
        else {
            final T ret = next.value;
            next.value = null;
            Unsafe.instance.putOrderedObject(this, tailOffset, next); //next will survive GC's if quiet
            return ret;
        }
    }
```

This can cause a lot of objects to be copied between survivor spaces (causing it to be very slow) and eventually will leak into tenured space.

The problem can be resolved by setting the tail to be a preallocated instance if the queue is empty. However, this means a few additional compareAndSet / Get and therefore has an impact on the performance of the queue. As it seems like its an edge case perhaps its acceptable in most cases (not in mine!)

Corresponding test to highlight the issue (you can see the tenured heap increasing massively) - use JVM options -Xms256m -Xmx256m -verbose:gc -server -XX:+UseParNewGC -XX:+UseConcMarkSweepGC -XX:MaxTenuringThreshold=4 and use a config that uses the SingleConsumerOnlyUnboundedMailbox

```
@Test
	public void testGCLeak() throws InterruptedException {
		ActorSystem akka = ActorSystem.create("test");
		ActorRef[] refs = new ActorRef[50000];
		for (int i = 0;i<refs.length;i++) {
			refs[i] = akka.actorOf(Props.create(DummyActor.class).withMailbox("default-mailbox")); //DummyActor does nothing
		}
		for (int i = 0;i<5;i++) {
			forceGC(); //force a minor GC 5 times
		}
		System.out.println("start");
		for (int y = 0;y<1000000;y++) {
			for (int i = 0;i<refs.length;i++) {
					refs[i].tell(new Object(), ActorRef.noSender());
			}
			for (int i = 0;i<refs.length;i++) {
				refs[i].tell(new Object(), ActorRef.noSender());
			}
			for (int i = 0;i<refs.length;i++) {
				refs[i].tell(new Object(), ActorRef.noSender());
			}
			Thread.sleep(2000);
			for (int i = 0;i<5;i++) {
				forceGC(); //force a minor GC 5 times
			}
			System.out.println("Iteration "+y);
		}
		
		System.out.println("done");
		
	}
	private void forceGC() throws InterruptedException {
		for (int i = 0;i<100000;i++) {
			new Object[1024].toString(); //encourage minor GC
		}
	}
```

I am looking for feedback on the correctness and possible optimizations on the fix I have applied to the class... As I have had to remove the lazySet of the node.next in favour of a CAS. And the consumer will be slower when the queue is empty/when consuming the last element of a queue.

Code changed in AbstractNodeQueue:
```
private final Node<T> tailInstance = new Node<T>(null);

protected GCFriendlyAbstractNodeQueue() {
		_tailDoNotCallMeDirectly = tailInstance;
		set(tailInstance);
}

@SuppressWarnings("restriction")
	public final T poll() {
		
		Node<T> tail = null;
		Node<T> current = null;
		Node<T> next = null;
		for(;;) {
			
			tail = ((Node<T>)Unsafe.instance.getObjectVolatile(this, tailOffset));
 			next = tail.next();
			current = get();
			if (next != null || current == tail)
				break;
		}

		if (next == null) 
			return null;
		else { 
			final T ret = next.value;
			next.value = null;

			if (current==next) { //no more elements
				Node<T> tailInstanceNext = tailInstance.next();
				if (this.compareAndSet(current,tailInstance)) { //if this passes, the current node is the root empty tail
					tailInstance.compareAndSetNext(tailInstanceNext, null); //this may fail but thats OK - either a producer has cleared the previous reference, or we are setting it to null
					Unsafe.instance.putOrderedObject(this, tailOffset, tailInstance); //ready for the next poll, we set the tail to the instance
				} else { //if it doesnt, a producer has added an element while consuming
					Unsafe.instance.putOrderedObject(this, tailOffset, next); //otherwise, a producer has already added to the linked list (the next node is the tail)
				}
			}
			else {
				Unsafe.instance.putOrderedObject(this, tailOffset, next);
			}
			return ret;
		}
	}

public static class Node<T> {
		public T value;
		@SuppressWarnings("unused")
		private final AtomicReference<Node<T>> _next = new AtomicReference<GCFriendlyAbstractNodeQueue.Node<T>>(null);

		public Node() {
			this(null);
		}

		public Node(final T value) {
			this.value = value;
		}

		@SuppressWarnings("unchecked")
		public final Node<T> next() {
			return _next.get();
		}

		protected final void setNext(final Node<T> newNext) {
			_next.set(newNext);
		}

		protected boolean compareAndSetNext(final Node<T> oldNode,final Node<T> newNode) {
			return _next.compareAndSet(oldNode,newNode);
		}

		@Override
		public String toString() {
			return super.toString()+"_next=" + _next + "]";
		}
		
		
	}


```

**Comment by drewhk** [comment]
It is not a leak though.

**Comment by dom54** [comment]
its not a memory leak. It is leaking garbage though no? In this scenario, some objects will only be collected in a full GC. In an optimised implementation they are collected in a minor GC.

**Comment by drewhk** [comment]
that is definitely a performance improvement, yes, but there is still no leak :)

**Comment by dom54** [comment]
Further optimization; reintroduced the lazySet on the Node value (no changes to that inner class now)
 now we can just compareAndSet(unusedInstance), then  lazySet(null) previous/now unused instance of tail.

Or am I missing something. I am sure there is a simpler way but I can't see it.

```

private final Node<T> tailInstance1 = new Node<T>(null);
private final Node<T> tailInstance2 = new Node<T>(null);
	
private volatile Node<T> tailInstance = tailInstance1;

@SuppressWarnings("restriction")
	public final T poll() {
		
		Node<T> tail = null;
		Node<T> current = null;
		Node<T> tailNext = null;
		for(;;) {
			
			tail = ((Node<T>)Unsafe.instance.getObjectVolatile(this, tailOffset));
 			tailNext = tail.next();
			current = get();
			if (tailNext != null || current == tail)
				break;
		}

		if (tailNext == null) 
			return null;
		else {
			final T ret = tailNext.value;
			tailNext.value = null;
			
			
			if (current==tailNext) { //no more elements, so reset the tail of the queue if possible to avoid a GC leak
				//however, a producer may add something to the queue while we are setting the tail element.
				//if that happens the "next" node will have a link to the new item. therefore we cannot perform the "reset"
				
				Node<T> previousTailInstance = tailInstance;
				Node<T> newTailInstance = previousTailInstance==tailInstance1?tailInstance2:tailInstance1; //swap to the new instance
				
				if (this.compareAndSet(current,newTailInstance)) { //if this passes, the current node is the root empty tail. newTailInstance is the one in use
					this.tailInstance = newTailInstance; //assign the current variable - volatile.
					previousTailInstance.setNext(null); //the previous one is guaranteed not to be in use as we have compareAndSet to the other instance
					Unsafe.instance.putOrderedObject(this, tailOffset, newTailInstance); //ready for the next poll, we set the tail to the instance (the non volatile one. as we know it cant change)
				} else { //if it doesnt, a producer has added an element while consuming.  we put the tailNext as the tail (traverse the linked list to next node)
					Unsafe.instance.putOrderedObject(this, tailOffset, tailNext); 
				}
			}
			else {
				Unsafe.instance.putOrderedObject(this, tailOffset, tailNext);
			}
			return ret;
		}
	}

```

**Comment by drewhk** [comment]
Can you open a PR instead? It is much easier to see differences there and add comments. Thank you!

**Comment by dom54** [comment]
Ok will give it a go...I'm new to Git so still finding my way around!

**Comment by viktorklang** [comment]
@dom54 Please also add a JMH bench so that performance can be compared. Thanks!

**Comment by huntc** [comment]
I *think* I'm feeling the pain of this right now also. Old gen GCs are slower than those for eden space. I'm currently suspecting that GC slowness is cause my Akka cluster based app to timeout its heartbeats. Perhaps @dom54 's suggestions here aren't just nice-to-have improvements.

Was there never a PR for this?

**Comment by dom54** [comment]
Nope ran out of time.  It's normal for old gen GC to be slow though. How many actors in your system?

**Comment by drewhk** [comment]
@huntc I don't think this affects the default mailbox, only the SingleConsumerOnlyMailbox. But @viktorklang or @rkuhn corrects me if I am wrong.

**Comment by viktorklang** [comment]
The LARS uses it too

-- 
Cheers,
√

**Comment by drewhk** [comment]
Ok that's true. But without a proper heap dump it is hard to be sure if this is a problem for @huntc. My initial gut hypothesis is that this is not relevant to his problems.

**Comment by viktorklang** [comment]
SelectionHandler ueses SerializedSuspendableExecutionContext which
uses AbstractNodeQueue

(I.e. IO related perhaps?)

On Wed, Dec 16, 2015 at 9:30 AM, drewhk <notifications@github.com> wrote:

> Ok that's true. But without a proper heap dump it is hard to be sure if
> this is a problem for @huntc <https://github.com/huntc>. My initial gut
> hypothesis is that this is not relevant to his problems.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/akka/akka/issues/17547#issuecomment-165030860>.
>



-- 
Cheers,
√

**Comment by drewhk** [comment]
A heap dump should tell. If there is a leak, then the leaked object will dominate the dump and it is easy to find.

**Comment by huntc** [comment]
Here's a heap dump for you @drewhk - I hope that it is useful to you - there are over 1 million `AbstractNodeQueue$Node` objects. I'm quite clueless as to the gradual increase of old gen that I've been observing - it may be entirely unrelated.

All that my program is doing is receiving a single http request via akka-http 1.0, replying, and then receiving the same request again. No requests are done in parallel. 

I really don't understand where all of these queue nodes come from - perhaps within akka streams/http 1.0?

[conductr.hprof.zip](https://github.com/akka/akka/files/63840/conductr.hprof.zip)

**Comment by drewhk** [comment]
@huntc sorry I won't debug your heap dump for you. Please follow what the AbstractNodeQueue nodes contain and backtrack it to the root cause. You first need to prove that your application is not overloading an internal queue to be able to prove it is a problem on the Akka side.

**Comment by viktorklang** [comment]
Unreachable objects:

There are 45k LARS$TaskQueues (AbstractNodeQueues)
There are 230k LARS$TaskHolders (TimeTasks)
There are 1.1m ANQ$Nodes

Are you doing something with a lot of timers?
Or may this be a result of the timers used by Http?

On Wed, Dec 16, 2015 at 11:24 AM, drewhk <notifications@github.com> wrote:

> @huntc <https://github.com/huntc> sorry I won't debug your heap dump for
> you. Please follow what the AbstractNodeQueue nodes contain and backtrack
> it to the root cause. You first need to prove that your application is not
> overloading an internal queue to be able to prove it is a problem on the
> Akka side.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/akka/akka/issues/17547#issuecomment-165058174>.
>



-- 
Cheers,
√

**Comment by huntc** [comment]
Endre and I also concluded that there were a lot of timers - my code is using `ask` from an http route. I shall avoid that and check out the difference. Thanks for looking.

**Comment by dom54** [comment]
FYI: Many ANQ$Node that get tenured but cleaned up by a full GC is the problem I was facing.

**Comment by viktorklang** [comment]
On Wed, Dec 16, 2015 at 12:00 PM, dom54 <notifications@github.com> wrote:

> FYI: Many ANQ$Node that get tenured but cleaned up by a full GC is the
> problem I was facing.
>
That is not surprising if there is a lot of task cancellations, since the
scheduler prunes the cancelled tasks when it visits those nodes.


> —
> Reply to this email directly or view it on GitHub
> <https://github.com/akka/akka/issues/17547#issuecomment-165066993>.
>



-- 
Cheers,
√

**Comment by huntc** [comment]
Related PR: https://github.com/akka/akka/pull/19220

**Comment by bantonsson** [comment]
Hi @dom54 as @huntc commented there is another issue than the "dangling" tail node being promoted, and that is that the ANQ suffered from neptosim, basically pulling in all nodes in the queue and all values into old space over time, since it didn't clear the next pointers of the nodes and the value pointer of discarded nodes. The ticket is #19216 

Getting rid of the tail node is a bit complicated as you have probably noticed by now. For example how do you make sure that the tailInstance is no longer in use in your second example? Once you have returned a node from `pollNode`, you have given up ownership and have no clue if it is being inserted into another ANQ with `addNode`, so you can't just reuse it.

**Comment by dom54** [comment]
Yep, so I removed pollNode :) its not required by Akka it seems, at least in the context of being a queue for an actor.

**Comment by bantonsson** [comment]
@dom54 I guess you haven't looked in `Scheduler.scala` :wink:

**Comment by bantonsson** [comment]
@dom54 I just did a quick run of your test (50 iterations) without and with the fixes in #19216, and these are the resulting graphs.
<img width="400" alt="akka 2 3 11 baseline" src="https://cloud.githubusercontent.com/assets/805830/11896218/def93af4-a584-11e5-80e4-393cac07c3d0.png"> <img width="400" alt="akka 2 3 11 with nepotism fixes" src="https://cloud.githubusercontent.com/assets/805830/11896226/eb4da83a-a584-11e5-94b0-ea3870c58124.png">

Quite a bit better I would say without really changing the queue.

**Comment by dom54** [comment]
Yep that does look better, will give it a go. The fix I eventually came up with is a bit different to the one above.   Regarding Schedulers.scala, what I meant was that in my case, I changed the default queue for the bulk of my own actors and it fixed the issue (Schedulers.scala still uses the existing implementation).  Changing the class across the board , as you rightly point out, is tricky with PollNode.

How do I generate those pretty graphs?

**Comment by bantonsson** [comment]
I used GCViewer https://github.com/chewiebug/GCViewer

**Comment by huafengw** [comment]
Hi all,
  We built a real-time streaming engine using Akka and when we were doing performance tuning, we met the same problem. Here is the related issue: https://github.com/gearpump/gearpump/issues/1816. Basically, our benchmark will launch about 100 Actors across 4 nodes and continually shuffles messages, the message size is 100 bytes. 

  We adopt the fix and built a new jar file to run our benchmark, and here is the result:
  **Before the fix:** The throughout of our benchmark is **14,600,000** msg/second
  **After the fix:** The throughput increase to **18,000,000** msg/second, also the receive latency reduced 1x.
  So we consider this fix to be very important to heavy-load Akka applications.

**Comment by viktorklang** [comment]
Hi @huafengw,

Fantastic verification, thank you!

**Comment by MQ-EL** [comment]
Hi @bantonsson ,  I think ```AbstractBoundedNodeQueue```  also suffers from nepotism

**Comment by bantonsson** [comment]
Hi @MQ-EL That's correct. It has the same issues as the `AbstractNodeQueue`. I did the changes in Akka 2.3, and missed that `AbstractBoundedNodeQueue` had been added in 2.4.

I've opened a ticket here #19614

**Comment by patriknw** [comment]
sounds like we have fixed this
